src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -54,16 +49,34 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
 
   let nextUnitOfWork : ?Fiber = null;
 
-  function performUnitOfWork(unit : Fiber) : ?Fiber {
-    switch (unit.tag) {
-      case FunctionalComponent:
-        return ReactFiberFunctionalComponent.performWork(unit);
-      case ClassComponent:
-        break;
-      case HostComponent:
-        break;
+  function completeUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    while (true) {
+      var next = completeWork(unitOfWork);
+      if (next) {
+        // If completing this work spawned new work, do that next.
+        return next;
+      } else if (unitOfWork.sibling) {
+        // If there is more work to do in this parent, do that next.
+        return unitOfWork.sibling;
+      } else if (unitOfWork.parent) {
+        // If there's no more work in this parent. Complete the parent.
+        unitOfWork = unitOfWork.parent;
+      } else {
+        // If we're at the root, there's no more work to do.
+        return null;
+      }
+    }
+  }
+
+  function performUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    var next = beginWork(unitOfWork);
```

Comment:
I'm not sure there is a current state because you can start a unit of work and then start the same thing again because of a higher priority operation or because you had to restart higher up before completing the first set.

The input is the next fiber that has work that needs to be done. The output is null if it completely immediately. If doing work on this fiber created children, more work is needed before we can complete the first one so we get that one as the next thing to begin work on.

Think of it this way:

```js
x = () => y(1 + 2) + 3;
```

The stack frame for executing x must first begin some initial work `1 + 2`. Then it jumps to `y`. Then it completes by finally performing `+ 3`. The stack frame of `x` is used twice.